File: src/com/suushiemaniac/cubing/alglib/util/ParseUtils.java
package com.suushiemaniac.cubing.alglib.util;

import com.suushiemaniac.cubing.alglib.alg.Algorithm;
import com.suushiemaniac.cubing.alglib.exception.InvalidNotationException;
import com.suushiemaniac.cubing.alglib.lang.*;

public class ParseUtils {
    public static boolean isParseable(String input, NotationReader parser) {
        try {
            Algorithm alg = parser.parse(input);
            return alg.length() >= 0;
        } catch (InvalidNotationException e) {
            return false;
        }
    }

    public static NotationReader getReaderForPuzzle(String puzzle) {
        switch (puzzle) {
            case "222":
            case "333":
            case "333ni":
            case "444":
            case "444ni":
            case "555":
            case "555ni":
            case "666":
            case "777":
                return new CubicAlgorithmReader();
            case "skewb":
                return new SkewbAlgorithmReader();
            case "pyram":
                return new PyraminxAlgorithmReader();
            case "minx":
                return new MegaminxAlgorithmReader();
            case "sq1":
                return new SquareOneAlgorithmReader();
            case "clock":
                return new ClockAlgorithmReader();
            default:
                return null;
        }
    }

    public static NotationReader guessReaderForAlgString(String algString) {
        if (algString.contains("/") || (algString.contains("(") && algString.contains(")")))
            return new SquareOneAlgorithmReader();
        else if (algString.contains("++") || algString.contains("--"))
            return new MegaminxAlgorithmReader();
        else if ((algString.contains("0") || algString.contains("1") || algString.contains("2") || algString.contains("3") || algString.contains("4") || algString.contains("5") || algString.contains("6")) && (algString.contains("+") || algString.contains("-")))
            return new ClockAlgorithmReader();
        else if (algString.contains("u") || algString.contains("d") || algString.contains("l") || algString.contains("r"))
            return new PyraminxAlgorithmReader();
        else if (algString.contains("D") || algString.contains("F"))
            return new CubicAlgorithmReader();
        else if (algString.contains("U") || algString.contains("B") || algString.contains("L") || algString.contains("R"))
            return new SkewbAlgorithmReader();
        else return null;
    }
}

File: src/main/kotlin/com/suushiemaniac/cubing/alglib/util/ParseUtils.kt
package com.suushiemaniac.cubing.alglib.util

import com.suushiemaniac.cubing.alglib.exception.InvalidNotationException
import com.suushiemaniac.cubing.alglib.lang.*
import com.suushiemaniac.cubing.alglib.move.NotationValue

object ParseUtils {
    fun <T : NotationValue> Array<T>.fromNotation(targetNotation: String): T {
        return this.find { it.notation == targetNotation }!!
    }

    fun isParseable(input: String, parser: NotationReader) = try {
        parser.parse(input).moveLength() >= 0
    } catch (e: InvalidNotationException) {
        false
    }

    fun getReaderForPuzzle(puzzle: String): NotationReader? = when (puzzle) {
        "222", "333", "333ni", "444", "444ni", "555", "555ni", "666", "777" -> CubicAlgorithmReader()
        "skewb" -> SkewbAlgorithmReader()
        "pyram" -> PyraminxAlgorithmReader()
        "minx" -> MegaminxAlgorithmReader()
        "sq1" -> SquareOneAlgorithmReader()
        "clock" -> ClockAlgorithmReader()
        else -> null
    }

    fun guessReaderForAlgString(algString: String): NotationReader? {
        if (algString.contains("/") || algString.contains("(") && algString.contains(")"))
            return SquareOneAlgorithmReader()
        else if (algString.contains("++") || algString.contains("--"))
            return MegaminxAlgorithmReader()
        else if ((algString.contains("0") || algString.contains("1") || algString.contains("2") || algString.contains("3") || algString.contains("4") || algString.contains("5") || algString.contains("6")) && (algString.contains("+") || algString.contains("-")))
            return ClockAlgorithmReader()
        else if (algString.contains("u") || algString.contains("d") || algString.contains("l") || algString.contains("r"))
            return PyraminxAlgorithmReader()
        else if (algString.contains("D") || algString.contains("F") || algString.contains("x") || algString.contains("y") || algString.contains("z"))
            return CubicAlgorithmReader()
        else if (algString.contains("U") || algString.contains("B") || algString.contains("L") || algString.contains("R"))
            return SkewbAlgorithmReader()
        else
            return null
    }
}
